pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,182 @@
+	ipAllocator.pendingIPRanges[ipRange] = true
+}
+
+// ReleaseIPRange releases the pending IPRange
+// Argument ipRange string is an IPV4 range which needs to be released
+func (ipAllocator *IPAllocator) ReleaseIPRange(ipRange string) {
+	ipAllocator.pendingIPRangesMutex.Lock()
+	defer ipAllocator.pendingIPRangesMutex.Unlock()
+	delete(ipAllocator.pendingIPRanges, ipRange)
+}
+
+// ValidateCIDR function validates whether a particular cidr is a valid IP range
+// Argument cidr string is is a CIDR range that needs to be validated
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) bool {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return false
+	}
+	// The cidr network id size needs to be maximum 29 bits
+	cidrSize, _ := ipnet.Mask.Size()
+	if cidrSize > maximumCIDRSize {
+		return false
+	}
+
+	mask := net.IPMask(net.ParseIP(ipRangeMask).To4())
+	// The IP specified in the reserved-ipv4-cidr must be the first IP in the CIDR block
+	return ip.String() == ip.Mask(mask).String()
+}
+
+// GetUnreservedIPRange returns an unreserved /29 IP block.
+// cidr: Provided cidr address in which we need to look for an unreserved /29 IP range
+// cloudInstancesReservedIPRanges: All the used IP ranges in the cloud instances
+// All the used IP ranges in the service instances not updated in cloud instances is extracted from the pendingIPRanges list in the IPAllocator
+// Finally a final reservedIPRange list is created by merging these two lists
+// Potential error cases:
+// 1) No /29 IP range in the CIDR is unreserved
+// 2) Parsing the CIDR resulted in an error
+func (ipAllocator *IPAllocator) GetUnreservedIPRange(cidr string, cloudInstancesReservedIPRanges map[string]bool) (string, error) {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", err
+	}
+	var reservedIPRanges = make(map[string]bool)
+
+	// The final reserved list is obtained by combining the cloudInstancesReservedIPRanges list and the pendingIPRanges list in the ipAllocator
+	for cloudInstancesReservedIPRange := range cloudInstancesReservedIPRanges {
+		reservedIPRanges[cloudInstancesReservedIPRange] = true
+	}
+
+	// Lock is placed here so that the pendingIPRanges list captures all the IPs pending reservation in the cloud instances
+	ipAllocator.pendingIPRangesMutex.Lock()
+	defer ipAllocator.pendingIPRangesMutex.Unlock()
+	for reservedIPRange := range ipAllocator.pendingIPRanges {
+		reservedIPRanges[reservedIPRange] = true
+	}
+
+	for cidrIPRange := cloneIP(ip.Mask(ipnet.Mask)); ipnet.Contains(cidrIPRange) && err == nil; cidrIPRange, err = incrementIP(cidrIPRange, incrementStep29IPRange) {
+		overLap := false
+		ipRange := fmt.Sprint(cidrIPRange.String(), reservationRange)
+		for reservedIPRange := range reservedIPRanges {
+			// Find if the current IP range in the CIDR overlaps with any of the reserved IP ranges. If not, this can be returned
+			overLap, err = isOverlap(ipRange, reservedIPRange)
```

Comment:
Can `isOverlap` take in ipnet as arguments?  ipRange was already an ipnet, and then it gets converted to a string, only to be converted back to an ipnet.

Reply:
It was not an IPnet, it was an IP. The concatenation converts it into a cidr which is then created into an IPnet

Reply:
Can you take the ip and use it to construct an ipnet by adding a mask?

Reply:
That would require masking and adding all 8 IPs manually to the slice

Reply:
Not sure I understand.  IPNet consists of a single IP + mask: https://golang.org/pkg/net/#IPNet